dig_test.go

```diff
@@ -1598,6 +1598,29 @@ func TestProvideCycleFails(t *testing.T) {
 			`depends on \*dig.D provided by "go.uber.org/dig".TestProvideCycleFails.\S+ \(\S+\)`,
 		)
 	})
+
+	t.Run("DeferAcyclicVerification bypasses cycle check, VerifyAcyclic catches cycle", func(t *testing.T) {
+		// A <- B <- C
+		// |         ^
+		// |_________|
+		type A struct{}
+		type B struct{}
+		type C struct{}
+		newA := func(*C) *A { return &A{} }
+		newB := func(*A) *B { return &B{} }
+		newC := func(*B) *C { return &C{} }
+
+		c := New(DeferAcyclicVerification())
+		assert.NoError(t, c.Provide(newA))
+		assert.NoError(t, c.Provide(newB))
+		assert.NoError(t, c.Provide(newC))
+
+		err := c.Invoke(func(*A) {})
+		require.Error(t, err, "expected error when introducing cycle")
+		assertErrorMatches(t, err,
+			`this function introduces a cycle:`,
```

Comment:
I think my concern with this approach is the error messaging. Since this test uses a regex, I'm not sure what the actual content of the message is.

But in essence, the reason why cycle verification is done on each node insertion is so you can pinpoint which element has caused the cycle. Are we still able to report the errors with the same accuracy? I think if the verification is deferred and the graph is complete (which could actually contain several cycles), are we able to properly point to the function (file, line number) that caused it?

Reply:
I have updated for clarity